### DISCUSSION TITLE
redis.clients.jedis.exceptions.JedisException: Could not get a resource from the pool

### DISCUSSION THREAD

**Discussion by SivaTharun2024** [xetorthio/jedis#4045] [Q&A]
I am seeing the below weird error, while using the jedis client to connect to the managed redis server instance, we have setup 10 different instances of the service, where we are using jedis client to connect to the redis instance. 

The issue is that, over time we have been noticing that one or more service instances, is not able to connect to the redis server and  we see the below error in the app logs.

`Connect error to redis server:10.xxx.xxx.x;Exception:redis.clients.jedis.exceptions.JedisException: Could not get a resource from the pool`

Only after restarting the effected app instance, the issue is getting resolved. i have checked the client info on the redis server, i have noticed that there are no issue on the available connections, nor there are any blocked clients.

<img width="248" alt="Screenshot 2024-12-30 at 9 43 41 AM" src="https://github.com/user-attachments/assets/88b19b80-99be-4114-b7d5-2e277ae018ce" />

Attaching the configuration code , on how we are setting up the jedis client for reference.

```
private static final RedisClient INSTANCE = new RedisClient();

 static {
        JedisPoolConfig config = new JedisPoolConfig();
        config.setMaxTotal(RedisClient.MAX_CONNECTIONS);
        config.setMaxIdle(RedisClient.MAX_IDLE_CONNECTIONS);
        config.setMaxWaitMillis(MAX_WAITINMILLIS);
        INSTANCE.init(config);
    }


  void init(JedisPoolConfig config) {
        setPoolConfig(config);
        setRedisHost(resolveRedisHost(RedisClient.DEFAULT_REDIS_HOST));
        initJedisPool(poolConfig, redisHost);
    }


 private void setPoolConfig(JedisPoolConfig jedisPoolConfig) {
        this.poolConfig = jedisPoolConfig;
    }

   private void setRedisHost(String redisHost) {
        this.redisHost = redisHost;
    }

 private void initJedisPool(JedisPoolConfig poolConfig, String redisHost) {
        this.connectionPool = new JedisPool(poolConfig, redisHost, Protocol.DEFAULT_PORT, RedisClient.TIMEOUT_MS);
    }

 
public Jedis getJedisClient() {
        try {
            return connectionPool.getResource();
        } catch (Exception e) {
            logRedisFailure(e);
            return null;
        }
    }
```

**Comment by sazzad16** [comment]
Firstly, please provide whole stacktrace.

**Reply by SivaTharun2024** [reply]
This is the stacktrace for your reference.



```
at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2078)
at org.apache.commons.pool2.impl.LinkedBlockingDeque.pollFirst(LinkedBlockingDeque.java:942)
       local variable: org.apache.commons.pool2.impl.LinkedBlockingDeque#1
  at org.apache.commons.pool2.impl.GenericObjectPool.borrowObject(GenericObjectPool.java:317)
       local variable: org.apache.commons.pool2.impl.GenericObjectPool#1
       local variable: java.time.Duration#36
  at org.apache.commons.pool2.impl.GenericObjectPool.borrowObject(GenericObjectPool.java:428)
  at org.apache.commons.pool2.impl.GenericObjectPool.borrowObject(GenericObjectPool.java:232)
  at redis.clients.util.Pool.getResource(Pool.java:49)
  at redis.clients.jedis.JedisPool.getResource(JedisPool.java:226)
       local variable: redis.clients.jedis.JedisPool#1
  at com.test.qi.rest.util.RedisClient.getJedisClient(RedisClient.java:118)
       local variable: com.test.qi.rest.util.RedisClient#1
  at com.test.qi.rest.util.RedisClient.getCount(RedisClient.java:199)

**Reply by sazzad16** [reply]
It can be seen that a resource (Jedis object) could not be received from the Pool. It is possible that all of MAX_CONNECTIONS resources in the Pool were busy and not enough resources were freed within MAX_WAITINMILLIS to serve the operation.

**Reply by SivaTharun2024** [reply]
This is the current pool configuration that we have right now on the redis, which is obtained via the `info` command on the redis cli. Do we need to explicitly set the `MAX_WAITINMILLIS `  when we configure the redis client here.

<img width="591" alt="Screenshot 2025-01-02 at 11 16 48 AM" src="https://github.com/user-attachments/assets/2b6a9edb-b26b-4cc1-8d0e-d6cc31f082d2" />

**Reply by sazzad16** [reply]
> Do we need to explicitly set the `MAX_WAITINMILLIS` when we configure the redis client here.

It depends on what do you want to achieve.

BTW, what do you actually mean? In your code, it shows `config.setMaxWaitMillis(MAX_WAITINMILLIS);`. So you are already configuring it. Right?

**Reply by SivaTharun2024** [reply]
Sorry, just overlooked this from my end, we are setting the MAX_WAITINMILLIS

**Comment by sazzad16** [comment]
Secondly, are you making sure if the Jedis object received from `getJedisClient()` is closed?

**Reply by SivaTharun2024** [reply]
Yes, we have defined `fetch`, `setKeyWithTTL` and `count` methods, to retrieve the relevant details for the corresponding keys, as below. Here we are making sure, that our client is closed whenever data insertion/retrieval operations are being performed.

```
public String fetch(String key) {
      
        String value = null;
        Jedis client = getJedisClient();
        if (client != null) {
            try {
                value = client.get(key);
                String data = "value=" + value;
            } catch (Exception e) {
                String data = "Failure=" + e.getMessage();
      			logFailure(e,data);
            } finally {
                client.close();
            }
        }
        return value;
    }

**Comment by sazzad16** [comment]
> ```
>         JedisPoolConfig config = new JedisPoolConfig();
>         config.setMaxTotal(RedisClient.MAX_CONNECTIONS);
>         config.setMaxIdle(RedisClient.MAX_IDLE_CONNECTIONS);
>         config.setMaxWaitMillis(MAX_WAITINMILLIS);
>         INSTANCE.init(config);
> ```

What are the actual values set to the `config` here?

**Reply by SivaTharun2024** [reply]
we are setting the below values in the config.

```
private static final int MAX_CONNECTIONS = 512;
private static final int MAX_IDLE_CONNECTIONS = 256;
private static final long MAX_WAITINMILLIS = 60000L; 
```

Do you see the necessity to set the MAX_WAITINMILLIS to more than 1 minute.  Also as i have attached in the earlier screenshot the value of the `maxclients` is at 65000, how will i get to know the value of the current active connections via the jedis pool.

**Reply by sazzad16** [reply]
> Do you see the necessity to set the MAX_WAITINMILLIS to more than 1 minute.

No.

**Reply by sazzad16** [reply]
For the number of connections to server, check `connected_clients`; which according to your previous screenshot is `2879`.

For the number of connections from the pool, check getNumActive() and getNumIdle().
You can also check getNumWaiters() to see how many are waiting. If this is too large, you may consider increasing maxIdle.

**Reply by mayankmaria** [reply]
Is there a way where we can re-start the Jedis client at Runtime.. i.e. based on ideal connection of Jedis.. if they are higher than some limit we would like to re-start (re-initialize) the entire Jedis pool once again. Is there any way to do the same ?

**Reply by sazzad16** [reply]
May be not. Perhaps you can re-initialize client object.
```java
    private void initJedisPool(JedisPoolConfig poolConfig, String redisHost) {
        // acquire lock
        this.connectionPool = new JedisPool(poolConfig, redisHost, Protocol.DEFAULT_PORT, RedisClient.TIMEOUT_MS);
        // release lock
    }
```
